platform/logger.cpp

```diff
@@ -0,0 +1,312 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#if defined(MBED_CONF_RTOS_PRESENT) && !defined(MBED_CONF_ZERO_BUFFER_LOGGING) && !defined(NDEBUG)
+#include <string.h>
+#include "platform/logger.h"
+#include "platform/CircularBuffer.h"
+#include "platform/mbed_wait_api.h"
+#include "hal/ticker_api.h"
+#include "hal/us_ticker_api.h"
+#include "platform/mbed_critical.h"
+#include "platform/mbed_interface.h"
+#include "rtos/EventFlags.h"
+
+#define LOG_MAX_BUFFER_SIZE_            (MBED_CONF_LOG_MAX_BUFFER_SIZE/sizeof(LOG_DATA_TYPE_))
+// Circular buffer should always have space for helper functions
+// in same line
+#define LOG_FREE_BUF_FLAG_              0x1u
+
+using namespace rtos;
+using namespace mbed;
+
+static CircularBuffer <LOG_DATA_TYPE_, LOG_MAX_BUFFER_SIZE_> log_buffer;
+static const ticker_data_t *const log_ticker = get_us_ticker_data();
+static uint32_t log_bytes_lost = 0;
+static EventFlags log_event_flag("logging");
+static LOG_DATA_TYPE_ *extern_buf = NULL;
+static uint32_t xbuf_count = 0;
+static bool time_enable = true;
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+void log_print_data(void)
```

Comment:
Where is declration for these functions? The line 43-45 would not needt o be here then if its part in the header file.